synapse/federation/transport/client.py

```diff
@@ -148,6 +157,46 @@ async def backfill(
             destination, path=path, args=args, try_trailing_slash_on_400=True
         )
 
+    @log_function
+    async def timestamp_to_event(
+        self, destination: str, room_id: str, timestamp: int, direction: str
+    ) -> TimestampToEventResponse:
+        """
+        Calls a remote federating server at `destination` asking for their
+        closest event to the given timestamp in the given direction.
+
+        Args:
+            destination: Domain name of the remote homeserver
+            room_id: Room to fetch the event from
+            timestamp: The point in time (inclusive) we should navigate from in
+                the given direction to find the closest event.
+            direction: ["f"|"b"] to indicate whether we should navigate forward
+                or backward from the given timestamp to find the closest event.
+
+        Returns:
+            Results in a dict received from the remote homeserver.
+            Expected response will include the `event_id` key for the closest event.
+
+        Raises:
+            Various exceptions when the request fails
+        """
+        path = _create_path(
+            FEDERATION_UNSTABLE_PREFIX,
+            "/org.matrix.msc3030/timestamp_to_event/%s",
+            room_id,
+        )
+
+        args = {"ts": [str(timestamp)], "dir": [direction]}
+
+        remote_response = await self.client.get_json(
+            destination, path=path, args=args, try_trailing_slash_on_400=True
+        )
+        assert isinstance(remote_response, dict)
+        assert remote_response.get("event_id", None) is not None
+        assert remote_response.get("origin_server_ts", None) is not None
```

Comment:
`assert` may not be the best tool for the job here. `assert`s are more for catching programmer mistakes rather than validation and are skipped entirely in `python -O` mode.

I took a look at how we do validation of federation responses. In `FederationClient` we raise `InvalidResponseError`s for malformed responses (see `get_room_hierarchy` in particular).

`InvalidResponseError` is defined in `federation_client.py`, which maybe suggests that's where validation should go? I'm not too clear on the distinction between `FederationClient` and `TransportLayerClient`.

We also need to check that `event_id` is a str and `origin_server_ts` is an int.
```py
        if not isinstance(remote_response, dict): ...
        if not isinstance(remote_response.get(...), str): ...
        if not isinstance(remote_response.get(...), int): ...
```

Reply:
It looks like some of the `TransportLayerClient` functions are cheating the types and just assuming it returns `JsonDict` when they could also return `List`

I've created a `timestamp_to_event` function in `federation_client.py` to do the rest of the validation like the other functions there ⏩